lib/TPP/GPU/LinalgToGpu.cpp

```diff
@@ -0,0 +1,149 @@
+//===- LinalgToGpu.cpp -------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Passes.h"
+
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+// Convert linalg.matmul to GPU-compatible kernel.
+struct ConvertMatmulToGpu : public OpRewritePattern<linalg::MatmulOp> {
+  using OpRewritePattern<linalg::MatmulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    if (!matmulOp.hasBufferSemantics())
+      return rewriter.notifyMatchFailure(matmulOp,
+                                         "Linalg to GPU expects memref type");
+
+    Location loc = matmulOp.getLoc();
+    ArrayRef<int64_t> shapeC = matmulOp.getDpsInitOperand(0)
+                                   ->get()
+                                   .getType()
+                                   .cast<ShapedType>()
+                                   .getShape();
+    ArrayRef<int64_t> shapeB =
+        matmulOp.getInputs()[1].getType().cast<ShapedType>().getShape();
+    ArrayRef<int64_t> shapeA =
+        matmulOp.getInputs()[0].getType().cast<ShapedType>().getShape();
+    if (shapeB.size() == 3) {
+      return rewriter.notifyMatchFailure(matmulOp,
+                                         "Packed BF16 loops unsupported");
+    }
+    // Parallel dims.
+    Value i = rewriter.create<arith::ConstantIndexOp>(loc, shapeC[0]);
+    Value j = rewriter.create<arith::ConstantIndexOp>(loc, shapeC[1]);
+    // Reduction dim.
+    Value k = rewriter.create<arith::ConstantIndexOp>(loc, shapeA[1]);
+    SmallVector<Value> ubs = {i, j, k};
+    // Lbs.
+    Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    SmallVector<Value> lbs = {zero, zero, zero};
+    // Step.
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    SmallVector<Value> steps = {one, one, one};
+    SmallVector<Value> ivs;
+
+    // Create parallel loops over the outer dimensions.
```

Comment:
This looks like packing, but for GPUs... We really need to use the same packing for all, and lower to GPUs after packing.

Reply:
Currently there is no packing involved in GPU lowering.
Linalg operation are only tiled to split them into GPU blocks operating on tiles which represent groups of threads per blocks.
Then this rewrite lowers linalg on tiles to elementwise operations such that they can be executed by individual threads.

Essentially, this rewrites `linalg.matmul` to a bit better kernel than the default `convert-linalg-to-parallel-loops` provides with respect to GPU.

Reply:
Hm, I see. This will play with the packing, then, we need these two to talk to each other later on.

Reply:
Depending on packing I'd expect need to generate different element-wise kernels. There will definitely be some need to coordinate the two but at the moment I have no examples. Although hopefully this last step will be replaced by some tile-based GPU kernel library like we do with libxsmm.